graphql_compiler/schema/schema_info.py

```diff
@@ -1,13 +1,166 @@
 # Copyright 2019-present Kensho Technologies, LLC.
 from collections import namedtuple
+from dataclasses import dataclass
+from functools import partial
+from typing import Dict, Optional
 
+from graphql.type import GraphQLSchema
 from graphql.type.definition import GraphQLInterfaceType, GraphQLObjectType
 import six
 import sqlalchemy
+from sqlalchemy.dialects.mssql import dialect as mssql_dialect
+from sqlalchemy.dialects.mysql import dialect as mysql_dialect
+from sqlalchemy.dialects.postgresql import dialect as postgresql_dialect
+from sqlalchemy.engine.interfaces import Dialect
 
 from . import is_vertex_field_name
 
 
+# Describes the intent to join two tables using the specified columns.
+#
+# The resulting join expression could be something like:
+# JOIN origin_table.from_column = destination_table.to_column
+#
+# The type of join (inner vs left, etc.) is not specified.
+# The tables are not specified.
+DirectJoinDescriptor = namedtuple(
+    "DirectJoinDescriptor",
+    (
+        "from_column",  # The column in the source table we intend to join on.
+        "to_column",  # The column in the destination table we intend to join on.
+    ),
+)
+
+
+@dataclass
+class GenericSchemaInfo:
```

Comment:
I don't think that using composition substantially improves usability. You know how to write `schema_info.generic_schema_info.schema` to get the schema and you definitely want to access the schema to see what it looks like.

Reply:
I think we can define a common base class for all the backend-specific schema info types that has properties for each of the `GenericSchemaInfo` fields, and hides the fact that the data actually lives one nesting level deeper.

Reply:
@pmantica1 makes a good point though I don't think I quite understand what you're proposing @obi1kenobi - would you mind providing an example?

Reply:
```python
@dataclass
class BackendSpecificSchemaInfo:
    generic_schema_info: GenericSchemaInfo

    @property
    def schema(self):
        return self.generic_schema_info.schema
    # etc for all properties of the generic schema 

@dataclass
class SQLSpecificSchemaInfo(BackendSpecificSchemaInfo):
    # SQL-specific fields here
    dialect: Dialect
    # etc for all other properties
```

Then, if `s = SQLSpecificSchemaInfo(...)`, the GraphQL schema is accessible as `s.schema` even though it's in reality stored as `s.generic_schema_info.schema`. This seems like the best of both worlds to me, though I'd love to hear if you have any concerns that I might have missed.

Reply:
If you run `get_sqlachemy_schema_info` you should get a `SQLSchemaInfo` not a `SQLSpecificSchemaInfo`  so you would still have to type: `s.sql_specific_info.schema`.

Reply:
@pmantica1 I'm not 100% sure I understand why this would be, but even if that is indeed the case, doesn't the same trick apply one more time? Add more property getter methods as necessary.

Reply:
Yes this trick sounds a good idea to me. What we have right now looks good in that sense. (I was trying to clarify that we should have `SQLSchemaInfo(BackendSpecificSchemaInfo)` which is what we have below. Also I want to make clear here that before `SQLSpecificSchemaInfo` was a field of `SQLSchemaInfo` so this was not only a naming issue)

Reply:
(Not in the scope of this diff) I'm starting to think that instead of a `schema info` object we might want to create an even more generic object per language that contains everything that language needs (including methods for compilation, etc). Thoughts?

Reply:
I think that would probably be unnecessary and net-negative. I'd like to distinguish between data (this schema info object) and computation (compilation functions etc.). With that split, there is no hidden state kept by the compilation functions other than what is accepted as input and what is produced as output.